lib/Runtime/Base/FunctionBody.cpp

```diff
@@ -8697,22 +8697,29 @@ namespace Js
                     // compact the types array by moving non-null types
                     // at the beginning.
                     this->types[nonNullIndex++] = type;
-                    isAnyTypeLive = true;
 #if DBG
                     isGuardValuePresent = this->guard->GetValue() == reinterpret_cast<intptr_t>(type) ? true : isGuardValuePresent;
 #endif
+                    this->types[i] = nullptr;
```

Comment:
i >= nonNullIndex. Carrying this->types[i] above should take care of the task since we already have `type` is looking to that index. Worst case i == nonNullIndex.